text/0317-third-party-dependency-management.md

```diff
@@ -0,0 +1,129 @@
+
+#### CLIs
+
+In our CLI packages (`aws-cdk`, `cdk-assets`, etc.) that are NodeJS-only,
+we have a very large dependency closure
+(around 180 packages - [see here](https://npmgraph.js.org/?q=aws-cdk)
+for the details).
+
+At build time of these packages,
+we generate an `npm-shrinkwrap.json` file with the exact version of each dependency
+(including transitive ones).
+This means a customer installing those packages in a given version will always get all of their dependencies in a specific version
+(the one defined in `npm-shrinkwrap.json`) as well.
+
+#### Framework
+
+In our framework packages, we have a much smaller dependency closure,
+roughly equal to:
+
+- `yaml@1.10.2`, which has 0 dependencies
+- `semver@7.3.5`, which depends on `lru-cache@6.0.0`, which depends on `yalist@4.0.0`
+- `jsonschema@1.4.0` (from `@aws-cdk/cloudformation-schema`)
+- `punycode@2.1.1` (from `@aws-cdk/aws-cognito`)
+- `case@1.6.3` (from `@aws-cdk/aws-codepipeline-actions`)
+- `@aws-cdk/core` depends on several packages for bundling: `ignore@5.1.8`, `@balena/dockerignore@1.0.2`, and:
+  - `minimatch@3.0.4`, which depends on `brace-expansion@1.1.11`, which depends on `balanced-match@1.02`, `concat-map@0.0.1`
+  - `fs-extra@9.1.0`, which depends on `at-least-node@1.0.0`, `jsonfile@6.1.0`, which depends on `universalify@2.0.0`, `graceful-fs@4.2.6`
+
+All of these third-party dependencies are bundled with the released framework packages.
+This is a JSII requirement
+(otherwise, these 3rd-party dependencies would not be available when using CDK from other languages!).
+
+For framework libraries in non-Node languages,
+the only extra dependencies
+(in addition to the bundled Node ones)
+are generally only the appropriate JSII runtime.
+These are updated separately, and are easier for customers to manage,
+as non-Node package managers make it easy to override the version of a transitive dependency
+like the JSII runtime in case a security vulnerability is found in it.
+Because of that, this RFC only focuses on managing Node dependencies.
+
+### Updating
+
+We have a GitHub Action that once a week updates all dependencies in the CDK repository to their latest,
+usually minor, version,
+using the [`npm-check-updates`](https://www.npmjs.com/package/npm-check-updates) package.
+
+### Summary
+
+Our current situation defends us pretty well against supply-chain attacks,
+because of the bundling that we do
+(our customers are always pinned to a specific version of a dependency).
+
+Also note that CDK is not susceptible to the attack described in the article
+["Researcher hacks over 35 tech firms in novel supply chain attack"](https://www.bleepingcomputer.com/news/security/researcher-hacks-over-35-tech-firms-in-novel-supply-chain-attack),
+as the CDK does not use any non-public packages.
+
+The disadvantage of bundling is that it doesn't make it easy for our customers
+to upgrade any third-party packages that have been found to have security vulnerabilities -
```

Comment:
Out of curiosity: do we run our own security scans on third party packages?

Reply:
I think indirectly we do via thinks like synk.io et al.

Reply:
The Amazon open source team might do that, but we (and by that "we", I mean the CDK team) do not run any additional scans.